examples/oauth/src/main.rs

```diff
@@ -0,0 +1,162 @@
+    AuthLayer, AuthUser, RequireAuthorizationLayer, SqliteStore,
+};
+use oauth2::{
+    basic::BasicClient, reqwest::async_http_client, AuthUrl, AuthorizationCode, ClientId,
+    ClientSecret, CsrfToken, RedirectUrl, Scope, TokenUrl,
+};
+use rand::Rng;
+use serde::Deserialize;
+use sqlx::{sqlite::SqlitePoolOptions, SqlitePool};
+
+#[derive(Debug, Default, Clone, sqlx::FromRow)]
+struct User {
+    id: i64,
+    password_hash: String,
+    name: String,
+}
+
+impl AuthUser for User {
+    fn get_id(&self) -> String {
+        format!("{}", self.id)
+    }
+
+    fn get_password_hash(&self) -> String {
+        self.password_hash.clone()
+    }
+}
+
+type AuthContext = axum_login::extractors::AuthContext<User, SqliteStore<User>>;
+
+#[tokio::main]
+async fn main() {
+    dotenvy::dotenv().ok();
+    let secret = rand::thread_rng().gen::<[u8; 64]>();
+
+    let session_store = MemoryStore::new();
+    let session_layer = SessionLayer::new(session_store, &secret).with_secure(false);
+
+    let db_url = env::var("DATABASE_URL").unwrap_or("oauth/user_store.db".to_string());
+
+    let pool = SqlitePoolOptions::new().connect(&db_url).await.unwrap();
+
+    let user_store = SqliteStore::<User>::new(pool.clone());
+    let auth_layer = AuthLayer::new(user_store, &secret);
+
+    let oauth_client = build_oauth_client();
+
+    let app = Router::new()
+        .route("/protected", get(protected_handler))
+        .route_layer(RequireAuthorizationLayer::<User>::login())
+        .route("/login", get(login_handler))
+        .route("/auth/google/callback", get(oauth_callback_handler))
+        .route("/logout", get(logout_handler))
+        .layer(Extension(oauth_client))
+        .layer(Extension(pool))
+        .layer(auth_layer)
+        .layer(session_layer);
+
+    axum::Server::bind(&"127.0.0.1:3000".parse().unwrap())
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+#[derive(Debug, Deserialize)]
+struct AuthRequest {
+    code: String,
+    state: String,
+}
+
+async fn oauth_callback_handler(
+    mut auth: AuthContext,
+    Query(query): Query<AuthRequest>,
+    Extension(pool): Extension<SqlitePool>,
+    Extension(oauth_client): Extension<BasicClient>,
+) -> impl IntoResponse {
+    // TODO: compare the crsf token generated before the request to the query.state
+
+    // Get an auth token
+    let _token = oauth_client
+        .exchange_code(AuthorizationCode::new(query.code))
+        .request_async(async_http_client)
+        .await
+        .unwrap();
+
+    // Do something with the token
+    // ...
+
+    // Fetch the user and log them in
+    let mut conn = pool.acquire().await.unwrap();
+    let user: User = sqlx::query_as("select * from users where id = 1")
+        .fetch_one(&mut conn)
+        .await
+        .unwrap();
+    auth.login(&user).await.unwrap();
+
+    println!("Logged in the user: {user:?}");
+
+    Redirect::to("/protected")
```

Comment:
This redirect results in a 401. It seems the login session is not preserved? Removing this redirect and then manually navigating to `/protected` does work, however.